### ISSUE TITLE
ver 4.35.2 transformers.Trainer, deepspeed backward(loss) is not used if distributed_state is NO

### ISSUE THREAD

**Issue by haixpham** [huggingface/transformers#27770]
### System Info

transformers 4.35.2
deepspeed 0.12.3
torch 2.1
CUDA 12.1

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

In 4.35.2 `Trainer.training_step()`:

```
if self.use_apex:
    with amp.scale_loss(loss, self.optimizer) as scaled_loss:
        scaled_loss.backward()
else:
    self.accelerator.backward(loss)
```
` self.accelerator.backward(loss)` only uses deepspeed engine's `backward(loss)` if distributed_state is Deepspeed, which is only set if launched with torchrun. If this is not the intended behavior, the fix is to either
- modify Trainer to detect this case
- modify Accelerator to handle this case in `backward()`

### Expected behavior

The below piece of code is from transformers 4.28.1, which always uses deepspeed when training_args.deepspeed is present.
```
if self.do_grad_scaling:
    self.scaler.scale(loss).backward()
elif self.use_apex:
    with amp.scale_loss(loss, self.optimizer) as scaled_loss:
        scaled_loss.backward()
elif self.deepspeed:
    # loss gets scaled under gradient_accumulation_steps in deepspeed
    loss = self.deepspeed.backward(loss)
else:
    loss.backward()

```

**Comment by ArthurZucker** [comment]
cc @muellerzr and @pacman100

**Comment by pacman100** [comment]
Hello, as per the documentation in https://huggingface.co/docs/transformers/main_classes/deepspeed#deployment-with-one-gpu and https://www.deepspeed.ai/getting-started/#launching-deepspeed-training, you need to use `deepspeed`, `torchrun` or `accelerate` launcher when using DeepSpeed. Just pass `--num_gpus=1` (with deepspeed launcher) or `nproc_per_node=1` (with torchrun launcher) or `num_processes=1` (with accelerate launcher).

**Comment by pacman100** [comment]
Hello, could you please let us know if the above PR resolves your issue?

**Comment by haixpham** [comment]
> Hello, could you please let us know if the above PR resolves your issue?

yes that would solve it at first glance. I will report back after running the whole training/eval process
